### PR TITLE
fix(router): throw an error when encounter undefined route

### LINKED ISSUE #12387: EXCEPTION: Uncaught (in promise): TypeError: Cannot read property 'visitExpression' of undefined
**I'm submitting a ...**  (check one with "x")

```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

```
error_handler.js:46 EXCEPTION: Uncaught (in promise): TypeError: Cannot read property 'visitExpression' of undefined
TypeError: Cannot read property 'visitExpression' of undefined
    at http://localhost:4000/vendor.bundle.js:14857:77
    at JitEmitterVisitor.AbstractEmitterVisitor.visitAllObjects (http://localhost:4000/vendor.bundle.js:14865:13)
    at JitEmitterVisitor.AbstractEmitterVisitor.visitAllExpressions (http://localhost:4000/vendor.bundle.js:14857:14)
    at JitEmitterVisitor.AbstractEmitterVisitor.visitLiteralArrayExpr (http://localhost:4000/vendor.bundle.js:14836:14)
    at LiteralArrayExpr.visitExpression (http://localhost:4000/vendor.bundle.js:15581:24)
    at http://localhost:4000/vendor.bundle.js:14857:78
    at JitEmitterVisitor.AbstractEmitterVisitor.visitAllObjects (http://localhost:4000/vendor.bundle.js:14865:13)
    at JitEmitterVisitor.AbstractEmitterVisitor.visitAllExpressions (http://localhost:4000/vendor.bundle.js:14857:14)
    at JitEmitterVisitor.AbstractEmitterVisitor.visitLiteralArrayExpr (http://localhost:4000/vendor.bundle.js:14836:14)
    at LiteralArrayExpr.visitExpression (http://localhost:4000/vendor.bundle.js:15581:24)
    at JitEmitterVisitor.AbstractEmitterVisitor.visitWritePropExpr (http://localhost:4000/vendor.bundle.js:14673:20)
    at WritePropExpr.visitExpression (http://localhost:4000/vendor.bundle.js:15383:24)
    at JitEmitterVisitor.AbstractEmitterVisitor.visitExpressionStmt (http://localhost:4000/vendor.bundle.js:14592:19)
    at ExpressionStatement.visitStatement (http://localhost:4000/vendor.bundle.js:15661:24)
    at http://localhost:4000/vendor.bundle.js:14873:58
    at Array.forEach (native)
    at JitEmitterVisitor.AbstractEmitterVisitor.visitAllStatements (http://localhost:4000/vendor.bundle.js:14873:20)
    at JitEmitterVisitor.AbstractEmitterVisitor.visitIfStmt (http://localhost:4000/vendor.bundle.js:14609:18)
    at IfStmt.visitStatement (http://localhost:4000/vendor.bundle.js:15748:24)
    at http://localhost:4000/vendor.bundle.js:14873:58
    at Array.forEach (native)
    at JitEmitterVisitor.AbstractEmitterVisitor.visitAllStatements (http://localhost:4000/vendor.bundle.js:14873:20)
    at JitEmitterVisitor.AbstractJsEmitterVisitor._visitClassGetter (http://localhost:4000/vendor.bundle.js:14973:18)
    at http://localhost:4000/vendor.bundle.js:14947:63
    at Array.forEach (native)
    at JitEmitterVisitor.AbstractJsEmitterVisitor.visitDeclareClassStmt (http://localhost:4000/vendor.bundle.js:14947:22)
    at ClassStmt.visitStatement (http://localhost:4000/vendor.bundle.js:15734:24)
    at http://localhost:4000/vendor.bundle.js:14873:58
    at Array.forEach (native)
    at JitEmitterVisitor.AbstractEmitterVisitor.visitAllStatements (http://localhost:4000/vendor.bundle.js:14873:20)
    at jitStatements (http://localhost:4000/vendor.bundle.js:16445:15)
    at RuntimeCompiler._compileModule (http://localhost:4000/vendor.bundle.js:17940:131)
    at RuntimeCompiler._compileModuleAndComponents (http://localhost:4000/vendor.bundle.js:17892:36)
    at RuntimeCompiler.compileModuleAsync (http://localhost:4000/vendor.bundle.js:17882:21)
    at ModuleBoundCompiler.compileModuleAsync (http://localhost:4000/vendor.bundle.js:18200:31)
    at MergeMapSubscriber.project (http://localhost:4000/vendor.bundle.js:51935:440)
    at MergeMapSubscriber._tryNext (http://localhost:4000/vendor.bundle.js:87167:27)
    at MergeMapSubscriber._next (http://localhost:4000/vendor.bundle.js:87157:18)
    at MergeMapSubscriber.Subscriber.next (http://localhost:4000/vendor.bundle.js:85209:18)
    at http://localhost:4000/vendor.bundle.js:86202:36
    at ZoneDelegate.invoke (http://localhost:4000/polyfills.bundle.js:15310:26)
    at Object.onInvoke (http://localhost:4000/vendor.bundle.js:36136:37)
    at ZoneDelegate.invoke (http://localhost:4000/polyfills.bundle.js:15309:32)
    at Zone.run (http://localhost:4000/polyfills.bundle.js:15192:43)
    at http://localhost:4000/polyfills.bundle.js:15580:57
    at ZoneDelegate.invokeTask (http://localhost:4000/polyfills.bundle.js:15343:35)
    at Object.onInvokeTask (http://localhost:4000/vendor.bundle.js:36127:37)
    at ZoneDelegate.invokeTask (http://localhost:4000/polyfills.bundle.js:15342:40)
    at Zone.runTask (http://localhost:4000/polyfills.bundle.js:15232:47)
    at drainMicroTaskQueue (http://localhost:4000/polyfills.bundle.js:15479:35)
```

**Expected behavior**
show page correctly.

**Minimal reproduction of the problem with instructions**
http://plnkr.co/edit/pjjwq8jJQHNUG0HK0Z7T?p=info

login component

**What is the motivation / use case for changing the behavior?**

**Please tell us about your environment:**
- **Angular version:** 2.0.0
- **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]
  
  
- **Language:** [all | TypeScript 2.X | ES6/7 | ES5]
- **Node (for AoT issues):** `node --version` =   5.0.0

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes https://github.com/angular/angular/issues/12387

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
It is in the @vsavkin domain but I would rather throw so people know that sth is wrong with their config. Plus this needs a test as without a test the filtering code could easily get deleted by accident.

**Comment by DzmitryShylovich** [comment]
Yeah, but all array methods (forEach, filter etc) skip values with `undefined` key :) 
You can test it 
```
var data = [42, 21,, 50, 40,, 9];
data.forEach( element ) => {
   console.log(element === undefined);
});

vs

var data = [42, 21, undefined, 50, 40, undefined, 9];
data.forEach( element ) => {
   console.log(element === undefined);
});
```

File: modules/@angular/router/src/router_module.ts [hunk 1]
```diff
@@ -194,6 +192,7 @@ export function provideForRootGuard(router: Router): any {
  * @stable
  */
 export function provideRoutes(routes: Routes): any {
+  validateConfig(routes);
```
Comment (vsavkin) [inline modules/@angular/router/src/router_module.ts]: Why do we need this call here? I think validateConfig should only be called in resetConfig.

Comment (DzmitryShylovich) [inline modules/@angular/router/src/router_module.ts]: I just wanted to reuse already existing functionality and fail fast with "descriptive" error msg

Comment (vsavkin) [inline modules/@angular/router/src/router_module.ts]: My point is that it will give you the same error, just a bit later.

**Comment by DzmitryShylovich** [comment]
@vsavkin updated

**Pull request merged**